File: schemas/cms-blog/drizzle/post_reactions.ts
// post_reactions: Emoji-style reactions on posts by authenticated users.
// See README.md for full design rationale.
import { pgTable, pgEnum, uuid, timestamp, unique, index } from "drizzle-orm/pg-core";
import { posts } from "./posts";

export const reactionTypeEnum = pgEnum("reaction_type", ["like", "love", "clap", "insightful", "bookmark"]);

export const postReactions = pgTable("post_reactions", {
  id: uuid("id").primaryKey().defaultRandom(),
  postId: uuid("post_id").notNull().references(() => posts.id, { onDelete: "cascade" }),
  userId: uuid("user_id").notNull(), // FK → users.id (cascade delete)
  reactionType: reactionTypeEnum("reaction_type").notNull(),
  createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
}, (table) => [
  unique("uq_post_reactions_post_id_user_id_reaction_type").on(table.postId, table.userId, table.reactionType),
  index("idx_post_reactions_user_id").on(table.userId),
  index("idx_post_reactions_post_id_reaction_type").on(table.postId, table.reactionType),
]);

File: schemas/cms-blog/spacetimedb/post_reactions.rs
// post_reactions: Emoji-style user reactions on posts.
// See README.md for full design rationale.
use spacetimedb::Timestamp;

#[derive(SpacetimeType, Clone)]
pub enum ReactionType {
    Like,       // type: String
    Love,
    Clap,
    Insightful,
    Bookmark,
}

#[spacetimedb::table(name = post_reactions, public)]
pub struct PostReaction {
    #[primary_key]
    pub id: String, // UUID
    #[index(btree)]
    pub post_id: String, // FK -> posts.id (cascade delete)
    #[index(btree)]
    pub user_id: String, // FK -> users.id (cascade delete)
    pub reaction_type: ReactionType, // Composite unique: (post_id, user_id, reaction_type)
    // Composite index: (post_id, reaction_type)
    pub created_at: Timestamp,
}
